hail/python/hailtop/aiotools/local_fs.py

```diff
@@ -205,14 +208,49 @@ async def remove(self, url: str) -> None:
         path = self._get_path(url)
         return await blocking_to_async(self._thread_pool, os.remove, path)
 
-    async def rmtree(self, sema: Optional[asyncio.Semaphore], url: str) -> None:
+    async def rmdir(self, url: str) -> None:
         path = self._get_path(url)
-
-        def f():
-            try:
-                shutil.rmtree(path)
-            except FileNotFoundError:
-                # no error if path does not exist
-                pass
-
-        await blocking_to_async(self._thread_pool, f)
+        return await blocking_to_async(self._thread_pool, os.rmdir, path)
+
+    async def rmtree(self,
+                     sema: Optional[asyncio.Semaphore],
+                     url: str,
+                     listener: Optional[Callable[[int], None]] = None) -> None:
+        if listener is None:
+            listener = lambda _: None  # noqa: E731
+        if sema is None:
+            sema = asyncio.Semaphore(50)
+
+        async def rm_file(path: str):
+            assert listener is not None
+            listener(1)
+            await self.remove(path)
+            listener(-1)
+
+        async def rm_dir(pool: OnlineBoundedGather2,
+                         contents_tasks: List[asyncio.Task],
+                         path: str):
+            assert listener is not None
+            listener(1)
+            if contents_tasks:
+                await pool.wait(contents_tasks)
+            await self.rmdir(path)
+            listener(-1)
+
+        async with OnlineBoundedGather2(sema) as pool:
+            tasks = []
+            rm_directory_task_by_url: Dict[str, asyncio.Task] = {}
+            for dirpath, dirnames, filenames in os.walk(url, topdown=False):
+                contents_tasks = [
+                    pool.call(rm_file, os.path.join(dirpath, filename))
+                    for filename in filenames
+                ] + [
```

Comment:
How does this work if you can't guarantee tasks will complete in a given order. i.e. I'm worried about the case where `rmdir(/tmp/foo)` finishes before `remove('/tmp/foo/bar')`?

Ok. I think I see what you are doing. You're waiting for one level at a time.